miniapps/navier/navier_solver.cpp

```diff
@@ -288,57 +292,80 @@ void NavierSolver::Setup(double dt)
    HInv->SetRelTol(rtol_hsolve);
    HInv->SetMaxIter(200);
 
+   // If the initial condition was set, it has to be aligned with dependent
+   // Vectors and GridFunctions
    un_gf.GetTrueDofs(un);
+   un_next = un;
+   un_next_gf.SetFromTrueDofs(un_next);
+
+   // Set initial time step in the history array
+   dthist[0] = dt;
 
    sw_setup.Stop();
 }
 
-void NavierSolver::Step(double &time, double dt, int cur_step)
+void NavierSolver::UpdateTimestepHistory(double dt)
+{
+   // Rotate values in time step history
+   dthist[2] = dthist[1];
+   dthist[1] = dthist[0];
+   dthist[0] = dt;
+
+   // Rotate values in nonlinear extrapolation history
+   Nunm2 = Nunm1;
+   Nunm1 = Nun;
+
+   // Rotate values in solution history
+   unm2 = unm1;
+   unm1 = un;
+
+   // Update the current solution and corresponding GridFunction
+   un_next_gf.GetTrueDofs(un_next);
+   un = un_next;
+   un_gf.SetFromTrueDofs(un);
+}
+
+void NavierSolver::Step(double &time, double dt, int cur_step, bool provisional)
 {
    sw_step.Start();
 
-   time += dt;
+   SetTimeIntegrationCoefficients(cur_step);
 
    // Set current time for velocity dirichlet boundary conditions.
    for (auto &vel_dbc : vel_dbcs)
    {
-      vel_dbc.coeff->SetTime(time);
+      vel_dbc.coeff->SetTime(time + dt);
    }
 
    // Set current time for pressure dirichlet boundary conditons.
    for (auto &pres_dbc : pres_dbcs)
    {
-      pres_dbc.coeff->SetTime(time);
+      pres_dbc.coeff->SetTime(time + dt);
    }
 
-   SetTimeIntegrationCoefficients(cur_step);
+   H_bdfcoeff.constant = bd0 / dt;
+   H_form->Update();
+   H_form->Assemble();
+   H_form->FormSystemMatrix(vel_ess_tdof, H);
 
-   if (cur_step <= 2)
+   if (partial_assembly)
    {
-      H_bdfcoeff.constant = bd0 / dt;
-      H_form->Update();
-      H_form->Assemble();
-      H_form->FormSystemMatrix(vel_ess_tdof, H);
-
-      if (partial_assembly)
-      {
-         HInv->SetOperator(*H);
-         delete HInvPC;
-         Vector diag_pa(vfes->GetTrueVSize());
-         H_form->AssembleDiagonal(diag_pa);
-         HInvPC = new OperatorJacobiSmoother(diag_pa, vel_ess_tdof);
-         HInv->SetPreconditioner(*HInvPC);
-      }
-      else
-      {
-         HInv->SetOperator(*H);
-      }
+      HInv->SetOperator(*H);
```

Comment:
Maybe nitpicking here, especially given that this structure is inherited from master, but could move this line before `if` and then remove `else`

Reply:
No you are right, the check is unnecessary now. Maybe I should check if the `dt` history is the same, but that could complicate things.